File: 338-counting-bits/counting-bits.py
class Solution:
    def countBits(self, n: int) -> List[int]:
        # ans=[]

        # for i in range(n+1):
        #     ans.append(str(bin(i)).count('1'))
        # return ans
        dp = [0]*(n+1)
        sub = 1
        for i in range(1,n+1):
            if sub*2==i:
                sub=i
            dp[i] = dp[i-sub] + 1
        return dp

File: 338-counting-bits/counting-bits.java
class Solution {
    public int[] countBits(int n) {
        int[] res = new int[n+1];
        res[0] = 0;
        int power = 2;
        for(int i=1;i<=n;i++){
            if(i%2==1)res[i]=res[i-1]+1;
            else if(i==power){
                res[i]=1;
                power*=2;
            }
            else
            res[i]=res[i-(power/2)]+1;
        }
        return res;
    }
}